pkg/kubelet/eviction/threshold_notifier_linux.go

```diff
@@ -67,51 +70,118 @@ func NewMemCGThresholdNotifier(path, attribute, threshold, description string, h
+	epfd, err := unix.EpollCreate1(0)
+	if err != nil {
+		glog.Warningf("eviction manager: error creating epoll: %v", err)
+		return nil, err
+	}
+	if epfd < 0 {
+		err = fmt.Errorf("EpollCreate1 call failed")
+		return nil, err
+	}
+	defer func() {
+		if err != nil {
+			unix.Close(epfd)
+		}
+	}()
+	config := fmt.Sprintf("%d %d %d", eventfd, watchfd, threshold)
 	_, err = unix.Write(controlfd, []byte(config))
 	if err != nil {
 		return nil, err
 	}
-	return &memcgThresholdNotifier{
-		watchfd:     watchfd,
-		controlfd:   controlfd,
-		eventfd:     eventfd,
-		handler:     handler,
-		description: description,
-	}, nil
+	notifier := &linuxCgroupNotifier{
+		watchfd:   watchfd,
+		controlfd: controlfd,
+		eventfd:   eventfd,
+		epfd:      epfd,
+		stop:      make(chan struct{}),
+	}
+	return notifier, nil
 }
 
-func getThresholdEvents(eventfd int, eventCh chan<- struct{}, stop ThresholdStopCh) {
+func (n *linuxCgroupNotifier) Start(eventCh chan<- struct{}) {
+	err := unix.EpollCtl(n.epfd, unix.EPOLL_CTL_ADD, n.eventfd, &unix.EpollEvent{
+		Fd:     int32(n.eventfd),
+		Events: unix.EPOLLIN,
+	})
+	if err != nil {
+		glog.Warningf("eviction manager: error adding epoll eventfd: %v", err)
+		return
+	}
 	for {
-		buf := make([]byte, 8)
-		_, err := unix.Read(eventfd, buf)
+		event, err := wait(n.epfd, n.eventfd)
 		if err != nil {
+			glog.Warningf("eviction manager: error while waiting for memcg events: %v", err)
 			return
+		} else if event {
+			// Consume the event from the eventfd
+			buf := make([]byte, 8)
+			_, err := unix.Read(n.eventfd, buf)
+			if err != nil {
+				glog.Warningf("eviction manager: error reading memcg events: %v", err)
+				return
+			}
+			select {
+			case <-n.stop:
+				return
+			case eventCh <- struct{}{}:
+			}
 		}
 
-		select {
-		case eventCh <- struct{}{}:
-		case <-stop.Ch():
+		if _, ok := <-n.stop; !ok {
 			return
 		}
 	}
 }
 
-func (n *memcgThresholdNotifier) Start(stop ThresholdStopCh) {
-	eventCh := make(chan struct{})
-	go getThresholdEvents(n.eventfd, eventCh, stop)
-	for {
-		select {
-		case <-stop.Ch():
-			glog.V(3).Infof("eviction: stopping threshold notifier")
-			unix.Close(n.watchfd)
-			unix.Close(n.controlfd)
-			unix.Close(n.eventfd)
-			return
-		case <-eventCh:
-			glog.V(2).Infof("eviction: threshold crossed")
-			n.handler(n.description)
+// wait waits up to notifierRefreshInterval for an event on the Epoll FD for the
+// eventfd we are concerned about.  It returns an error if one occurrs, and true
+// if the consumer should read from the eventfd.
+func wait(epfd, eventfd int) (bool, error) {
+	events := make([]unix.EpollEvent, 7)
+	timeoutMS := int(notifierRefreshInterval / time.Millisecond)
+	n, err := unix.EpollWait(epfd, events, timeoutMS)
+	if n == -1 {
+		if err == unix.EINTR {
+			// Interrupt, ignore the error
+			return false, nil
 		}
+		return false, err
+	}
+	if n == 0 {
+		// Timeout
+		return false, nil
+	}
+	if n > 6 {
```

Comment:
Please add a comment to explain why 6 is the magic number here...

You might want to declare a function, since you also use 6+1 when creating the events slice.

Reply:
added a constant.